fix 23355: invalid template parameter loses error location in some cases (T[UndefinedIdentifier] has no error loc)

WebFreak001 opened the pull request:
Types (e.g. TypeAArray) don't have a location attached to them, so instead of trying to use that we instead use the error location from the general template parameters.

@maxhaton 

The error message came from here: https://github.com/dlang/dmd/blob/58a63ffcedaada9b7501aea60d02a28b12c8770a/compiler/src/dmd/typesem.d#L175
which was called from the here changed file.

maxhaton commented on compiler/src/dmd/dtemplate.d:
```diff
@@ -3841,10 +3841,20 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
                         Type tx;
                         Dsymbol s;
-                        taa.index.resolve(Loc.initial, sc, e, tx, s);
+                        taa.index.resolve(loc, sc, e, tx, s);
```
Mine is less wrong than yours. The thing that actually fails is a function parameter rather than a template parameter though.

WebFreak001 replied:
I'm afraid there is some exotic way to get parameters to have 0 parameters here, which would break with this code.

WebFreak001 replied:
inferStart seems to never be used as a regular index in other code in this file, but there have been multiple other ways how parameters is indexed. My original code here copies from above: https://github.com/dlang/dmd/blob/58a63ffcedaada9b7501aea60d02a28b12c8770a/compiler/src/dmd/dtemplate.d#L3538

maxhaton replied:
Unless there is a non-zero number of template parameters, one of which is called `T`, this code will never run.

maxhaton commented on compiler/test/fail_compilation/diag23355.d:
```diff
@@ -0,0 +1,11 @@
+---
+*/
+void ffi(T)(T[n] s) { }
+
```
0 template parameter test
```suggestion
void ffi()(T[n] s)
{
}
```

WebFreak001 replied:
that one is unrelated to this fix, already worked before

maxhaton replied:
this is in reference to your point about parameters being of zero length.

WebFreak001 replied:
ok added and using inferStart now

maxhaton replied:
![image](https://user-images.githubusercontent.com/5672948/191619169-449181c0-c956-4d55-a55d-95ade91ab398.png)
gib coauthor

WebFreak001 replied:
done :p

maxhaton commented on compiler/src/dmd/dtemplate.d:
```diff
@@ -3841,10 +3841,20 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
                     else
                     {
+                        Loc loc;
+                        // use the first location of template parameters as we
```
```
The "type" (it hasn't been resolved yet) of the function parameter
does not have a location but the parameter it is related to does, so we use
that for the resolution (better error message).
```
is the comment I wrote on my patch

WebFreak001 committed:
fix 23355: T[UndefinedIdentifier] has no error loc

Types (e.g. TypeAArray) don't have a location attached to them, so
instead of trying to use that we instead
use the error location from the general template parameters.

Co-authored-by: mhh <maxhaton@gmail.com>

maxhaton commented on compiler/src/dmd/dtemplate.d:
```diff
@@ -3841,10 +3841,20 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
+                        // use the first location of template parameters as we
+                        // don't have any other location set anywhere.
+                        // not completely accurate, but better than nothing.
+                        if (inferStart < parameters.dim)
```
can this ever be false?

WebFreak001 replied:
inferStart has a few assignments in a few places in the code where it's assigned to `parameters.dim`, would prefer to keep this check because I have not analyzed all the different sources where this function may be called from.

maxhaton replied:
It might be right to have it be an assertion.

Those are set to make [inferStart..end] == [] I think.

Issue is whether there's going to be an assertion failure because it. These functions are the back-rooms of dmd

WebFreak001 replied:
I think this whole file really needs a refactoring, there is a lot of code duplication happening here.

maxhaton replied:
That need for refactoring is why we don't have named args
